orchagent/dash/dashaclgroupmgr.cpp

```diff
@@ -0,0 +1,721 @@
+    {
+        SWSS_LOG_INFO("Update full ACL group %s", group_id.c_str());
+
+        refreshAclGroupFull(group_id);
+    }
+    else
+    {
+        SWSS_LOG_INFO("Update ACL group %s", group_id.c_str());
+        for (auto& rule_it: group.m_dash_acl_rule_table)
+        {
+            const auto& rule_id = rule_it.first;
+            auto& rule = rule_it.second;
+            if (rule.m_src_tags.find(tag_id) != rule.m_src_tags.end() || rule.m_dst_tags.find(tag_id) != rule.m_dst_tags.end())
+            {
+                auto status = removeRule(group, rule);
+                if (status != SAI_STATUS_SUCCESS)
+                {
+                    SWSS_LOG_INFO("Failed to update ACL tag %s for ACL rule %s:%s, Failed to remove old rule, rv: %s", 
+                        tag_id.c_str(), group_id.c_str(), rule_id.c_str(), sai_serialize_status(status).c_str());
+                    continue;
+                }
+
+                status = createRule(group, rule);
+                if (status != SAI_STATUS_SUCCESS)
+                {
+                    SWSS_LOG_INFO("Failed to update ACL tag %s for ACL rule %s:%s, Failed to create  rule, rv: %s", 
+                        tag_id.c_str(), group_id.c_str(), rule_id.c_str(), sai_serialize_status(status).c_str());
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+task_process_status DashAclGroupMgr::refreshAclGroupFull(const string &group_id)
+{
+    SWSS_LOG_ENTER();
+
+    if (!exists(group_id))
+    {
+        return task_failed;
+    }
+
+    auto& group = m_groups_table[group_id];
+
+    DashAclGroup new_group = group;
+    init(new_group);
+    auto status = create(new_group);
+    if (status != task_success)
+    {
+        SWSS_LOG_INFO("Failed to perform full ACL group update %s. Failed to create group, rv: %s", group_id.c_str(), sai_serialize_status(status).c_str());
+        return task_failed;
+    }
+
+    for (auto& rule: new_group.m_dash_acl_rule_table)
+    {
+        const auto &rule_id = rule.first;
+        auto status = createRule(new_group, rule.second);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to perform full ACL group update %s. Failed to create DASH ACL rule %s, rv: %s", 
+                group_id.c_str(), rule_id.c_str(), sai_serialize_status(status).c_str());
+            
+            removeAclGroupFull(new_group);
+            return task_failed;
+        }
+    }
+
+    for (const auto& table: new_group.m_in_tables)
+    {
+        const auto& eni_id = table.first;
+        const auto& stages = table.second;
+
+        const auto eni = m_dash_orch->getEni(eni_id);
+        ABORT_IF_NOT(eni != nullptr, "Failed to get ENI %s", eni_id.c_str());
+
+        for (const auto& stage: stages)
+        {
+            auto status = bind(new_group, *eni, DashAclDirection::IN, stage);
+            ABORT_IF_NOT(status == SAI_STATUS_SUCCESS, "Failed to bind new group to ENI");
```

Comment:
It is a good question. If an error happens in the middle of the ACL group refreshing during binding of the new group to the ENI this means that something is wrong with the system. We can't recover from a situation like that because it is unknown what to do. If the binding of the new group failed we are not sure in which state that stage of the ENI is left. Does the old group is still bound? If we try to bind the old group will it fail? If we recover the old group in which state is the binding of the other ENIs? Do we need to back up to the old group?
In this situation asserting is the right way to handle the error. Orchagent will crash and restart. The controller will detect this and really the configuration

Reply:
@dgsudharsan done